File: apps/web/src/mocks/handlers/research.ts
import { HttpResponse, http, type RequestHandler } from "msw";
import { mockExperimentSummaryList } from "../fixtures/experiment-live";
import {
	mockExperimentArtifacts,
	mockExperimentComparison,
	mockExperimentDetail,
	mockExperimentGates,
	mockExperimentSelectionEvidence,
} from "../fixtures/experiment-workbench";
import { mockFactorCatalog } from "../fixtures/factor-catalog-live";
import {
	mockExperiments,
	mockFactorAnalysis,
	mockFactorDetail,
	mockFactors,
	mockReviewQueue as mockPrototypeReviewQueue,
	mockResearchPulse,
	mockResearchRuns,
} from "../fixtures/research";
import { mockReviewPacket, mockReviewQueue } from "../fixtures/review-live";

export const researchHandlers: RequestHandler[] = [
	http.get("/api/research/pulse", () => {
		return HttpResponse.json(mockResearchPulse);
	}),

	http.get("/api/factors", ({ request }) => {
		const url = new URL(request.url);
		const page = Number(url.searchParams.get("page") ?? 1);
		const pageSize = Number(url.searchParams.get("pageSize") ?? 20);
		const start = (page - 1) * pageSize;
		const items = mockFactors.items.slice(start, start + pageSize);

		return HttpResponse.json({
			items,
			total: mockFactors.total,
			page,
			pageSize,
		});
	}),

	http.get("/api/factors/:id", () => {
		return HttpResponse.json(mockFactorDetail);
	}),

	http.get("/api/factors/:id/analysis", () => {
		return HttpResponse.json(mockFactorAnalysis);
	}),

	http.get("/api/research/runs", () => {
		return HttpResponse.json(mockResearchRuns);
	}),

	http.get("/api/research/experiments", () => {
		return HttpResponse.json(mockExperiments);
	}),

	http.get("/api/research/review-queue", () => {
		return HttpResponse.json(mockPrototypeReviewQueue);
	}),

	// === R3 live-shape handler（/api/v1/research/experiments，generated DTO + {data} 信封）===
	http.get("/api/v1/research/experiments", () =>
		HttpResponse.json({
			data: mockExperimentSummaryList,
			pagination: { total: mockExperimentSummaryList.length, limit: 50, offset: 0, has_more: false },
		}),
	),
	http.post("/api/v1/research/experiments/:experimentId/preflight", () =>
		HttpResponse.json({
			data: {
				status: "ready",
				plan_hash: "d".repeat(64),
				checks: [],
				candidate_count: 3,
				planned_fold_count: 12,
				budget_run_count: 12,
				estimated_trading_sessions: 2048,
				estimated_disk_bytes: 4096,
				eligible_month_count: 96,
				isolation_width_sessions: 6,
			},
		}),
	),
	http.post("/api/v1/research/experiments", async ({ request }) => {
		const body = (await request.json()) as { experiment_id?: string; confirmed_plan_hash?: string };
		return HttpResponse.json({
			data: {
				experiment_id: body.experiment_id ?? "r3-experiment",
				status: "queued",
				queue_ordinal: 1,
				revision: 1,
				plan_hash: body.confirmed_plan_hash ?? "d".repeat(64),
				candidate_count: 3,
				fold_count: 12,
			},
		});
	}),
	http.get("/api/v1/research/experiments/:experimentId", () => HttpResponse.json({ data: mockExperimentDetail })),
	http.get("/api/v1/research/experiments/:experimentId/candidates", () =>
		HttpResponse.json({ data: mockExperimentDetail.candidates }),
	),
	http.get("/api/v1/research/experiments/:experimentId/gates", () => HttpResponse.json({ data: mockExperimentGates })),
	http.get("/api/v1/research/experiments/:experimentId/comparison", () =>
		HttpResponse.json({ data: mockExperimentComparison }),
	),
	http.get("/api/v1/research/experiments/:experimentId/artifacts", () =>
		HttpResponse.json({ data: mockExperimentArtifacts }),
	),
	http.get("/api/v1/research/experiments/:experimentId/selection-evidence", () =>
		HttpResponse.json({ data: mockExperimentSelectionEvidence }),
	),
	http.post("/api/v1/research/experiments/:experimentId/:action", async ({ params, request }) => {
		if (!new Set(["pause", "cancel", "resume", "retry-fold"]).has(String(params["action"]))) return undefined;
		const body = (await request.json()) as { expected_revision?: number };
		return HttpResponse.json({
			data: {
				experiment_id: String(params["experimentId"]),
				status: String(params["action"]),
				desired_state: String(params["action"]),
				revision: (body.expected_revision ?? mockExperimentDetail.revision) + 1,
				live_run_ids: [],
				occurred_at: "2026-08-01T00:00:00Z",
			},
		});
	}),
	http.post("/api/v1/research/experiments/:experimentId/candidate-selection", async ({ params, request }) => {
		const body = (await request.json()) as {
			candidate_id: string;
			comparison_payload_hash: string;
			expected_revision: number;
		};
		return HttpResponse.json({
			data: {
				selection_id: "selection-mock-1",
				experiment_id: String(params["experimentId"]),
				candidate_id: body.candidate_id,
				comparison_payload_hash: body.comparison_payload_hash,
				candidate_evidence_artifact_id: "candidate-bundle-2",
				candidate_evidence_content_hash: "2".repeat(64),
				selection_evidence_content_hash: mockExperimentSelectionEvidence.content_hash,
				revision: body.expected_revision + 1,
				event_id: "event-selection-mock-1",
				occurred_at: "2026-08-01T00:00:00Z",
			},
		});
	}),
	http.post("/api/v1/research/experiments/:experimentId/holdout-evaluations", async ({ params, request }) => {
		const body = (await request.json()) as {
			candidate_id: string;
			selection_id: string;
			expected_revision: number;
			expected_candidate_evidence_content_hash: string;
			expected_selection_evidence_hash: string;
		};
		return HttpResponse.json({
			data: {
				selection_id: body.selection_id,
				claim_id: "claim-mock-1",
				experiment_id: String(params["experimentId"]),
				candidate_id: body.candidate_id,
				fold_id: "holdout",
				logical_run_id: "run-holdout-mock-1",
				reproduction_fingerprint: "9".repeat(64),
				candidate_evidence_content_hash: body.expected_candidate_evidence_content_hash,
				selection_evidence_content_hash: body.expected_selection_evidence_hash,
				claim_payload_hash: "8".repeat(64),
				revision: body.expected_revision + 1,
				event_id: "event-holdout-mock-1",
				occurred_at: "2026-08-01T00:01:00Z",
				state: "claimed",
			},
		});
	}),
	http.get("/api/v1/research/candidates/:candidateId/selections", ({ params, request }) => {
		const experimentId = new URL(request.url).searchParams.get("experiment_id") ?? "exp-1042";
		return HttpResponse.json({
			data: {
				candidate_id: String(params["candidateId"]),
				experiment_id: experimentId,
				artifact_id: "candidate-bundle-2",
				content_hash: "2".repeat(64),
				items: [
					{
						fold_id: "fold-1",
						validation_fold_ordinal: 1,
						trade_date: "2021-01-04",
						instrument_id: "000001",
						rank: 1,
						score: 0.9,
						selected: true,
						evidence_hash: "a".repeat(64),
					},
				],
				next_cursor: null,
			},
		});
	}),
	http.get("/api/v1/research/candidates/:candidateId/exclusions", ({ params, request }) => {
		const experimentId = new URL(request.url).searchParams.get("experiment_id") ?? "exp-1042";
		return HttpResponse.json({
			data: {
				candidate_id: String(params["candidateId"]),
				experiment_id: experimentId,
				artifact_id: "candidate-bundle-2",
				content_hash: "2".repeat(64),
				items: [],
				next_cursor: null,
			},
		});
	}),
	http.get("/api/v1/research/candidates/:candidateId/factor-contributions", ({ params, request }) => {
		const experimentId = new URL(request.url).searchParams.get("experiment_id") ?? "exp-1042";
		return HttpResponse.json({
			data: {
				candidate_id: String(params["candidateId"]),
				experiment_id: experimentId,
				artifact_id: "candidate-bundle-2",
				content_hash: "2".repeat(64),
				items: [
					{
						fold_id: "fold-1",
						validation_fold_ordinal: 1,
						trade_date: "2021-01-04",
						instrument_id: "000001",
						factor_id: "momentum",
						rank: 1,
						selected: true,
						contribution: 0.12,
						evidence_hash: "c".repeat(64),
					},
				],
				next_cursor: null,
			},
		});
	}),

	http.get("/api/v1/research/factors/:factorId/diagnostics", ({ params, request }) => {
		const query = new URL(request.url).searchParams;
		return HttpResponse.json({
			data: {
				factor_id: String(params["factorId"]),
				snapshot_id: query.get("snapshot_id") ?? "snapshot-r3",
				snapshot_hash: "a".repeat(64),
				registry_hash: query.get("registry_hash") ?? "f".repeat(64),
				start_date: query.get("start_date") ?? "2024-01-01",
				end_date: query.get("end_date") ?? "2024-12-31",
				provenance: { dataset_id: "stock_daily" },
				metrics: { coverage: 0.97, rank_ic: 0.08 },
				artifact_id: "factor-diagnostic-1",
				content_hash: "c".repeat(64),
			},
		});
	}),
	http.get("/api/v1/research/factors", () => HttpResponse.json({ data: mockFactorCatalog })),

	// === R3 review queue + review-packet live-shape（generated DTO + {data} 信封）===
	http.get("/api/v1/research/reviews", () => HttpResponse.json({ data: mockReviewQueue })),
	http.get("/api/v1/research/experiments/:experimentId/review-packet", ({ params }) => {
		const experimentId = String(params["experimentId"]);
		if (experimentId !== mockReviewPacket.experiment_id) {
			return new HttpResponse(null, { status: 404 });
		}
		return HttpResponse.json({ data: mockReviewPacket });
	}),
];

File: packages/apps/src/ditto_apps/models/research.py
"""Research experiment API DTOs."""

from __future__ import annotations

from datetime import date, datetime
from typing import Any, Literal

from ditto_platform.foundation.json_types import JsonValue
from pydantic import BaseModel, ConfigDict

__all__ = [
    "CandidateExclusionEventResponse",
    "CandidateExclusionPageResponse",
    "CandidateFactorContributionPageResponse",
    "CandidateFactorContributionResponse",
    "CandidateSelectionEventResponse",
    "CandidateSelectionPageResponse",
    "CandidateSelectionReceiptResponse",
    "CandidateSelectionRequest",
    "ExperimentArtifactResponse",
    "ExperimentCandidateResponse",
    "ExperimentComparisonResponse",
    "ExperimentControlReceiptResponse",
    "ExperimentControlRequest",
    "ExperimentDetailResponse",
    "ExperimentFoldResponse",
    "ExperimentGateResponse",
    "ExperimentLaunchRequest",
    "ExperimentLaunchResponse",
    "ExperimentPlanningRequest",
    "ExperimentPreflightCheckResponse",
    "ExperimentPreflightResponse",
    "ExperimentRetryFoldRequest",
    "ExperimentReviewPacketResponse",
    "ExperimentSelectionEvidenceResponse",
    "ExperimentSummaryResponse",
    "FactorDescriptorResponse",
    "FactorDiagnosticsResponse",
    "HoldoutEvaluationReceiptResponse",
    "HoldoutEvaluationRequest",
    "NodeDescriptorResponse",
    "ReviewExposureWeightResponse",
    "ReviewGateOutcomeResponse",
    "ReviewSelectionExposureResponse",
    "ReviewSelectionTraceRefResponse",
]


class _StrictPlanningModel(BaseModel):
    """Fail-closed transport base for one canonical planning document node."""

    model_config = ConfigDict(
        strict=True,
        frozen=True,
        extra="forbid",
        allow_inf_nan=False,
    )


class ExperimentPlanningStrategyRequest(_StrictPlanningModel):
    """Exact strategy-version content identity supplied to planning."""

    strategy_id: str
    version: int
    spec_hash: str
    spec_json: dict[str, JsonValue]


class ExperimentPlanningSnapshotRequest(_StrictPlanningModel):
    """Exact certified research snapshot identity supplied to planning."""

    snapshot_id: str
    manifest_hash: str


class ExperimentPlanningBaselineRequest(_StrictPlanningModel):
    """Typed baseline descriptor envelope for the candidate matrix."""

    descriptor_type: str
    payload: dict[str, JsonValue]
    schema_version: int


class ExperimentPlanningParameterValueRequest(_StrictPlanningModel):
    """One tagged scalar matrix value."""

    type: Literal["bool", "int", "float", "string"]
    value: bool | int | float | str


class ExperimentPlanningParameterAxisRequest(_StrictPlanningModel):
    """One ordered candidate-matrix parameter axis."""

    name: str
    values: list[ExperimentPlanningParameterValueRequest]


class ExperimentPlanningMatrixRequest(_StrictPlanningModel):
    """Complete candidate matrix preimage."""

    baseline: ExperimentPlanningBaselineRequest
    axes: list[ExperimentPlanningParameterAxisRequest]
    candidate_limit: int


class ExperimentPlanningDatasetRequirementRequest(_StrictPlanningModel):
    """One exact certified dataset binding."""

    dataset_id: str
    expected_snapshot_ids: list[str]
    requires_pit_universe: bool
    certified_from: str | None


class ExperimentPlanningCostModelRequest(_StrictPlanningModel):
    """Deterministic resource-estimation coefficients."""

    bytes_per_run: int
    bytes_per_trading_session: int


class ExperimentPlanningBudgetRequest(_StrictPlanningModel):
    """Pre-registered hard resource ceilings."""

    candidate_limit: int
    fold_run_limit: int
    trading_session_limit: int
    disk_byte_limit: int


class ExperimentPlanningRequest(_StrictPlanningModel):
    """Complete transport-only document passed unchanged to the builder."""

    experiment_id: str
    research_cycle_id: str
    research_cycle_hash: str
    strategy: ExperimentPlanningStrategyRequest
    snapshot: ExperimentPlanningSnapshotRequest
    validation: dict[str, JsonValue]
    matrix: ExperimentPlanningMatrixRequest
    promotion_objective: dict[str, JsonValue]
    dataset_requirements: list[ExperimentPlanningDatasetRequirementRequest]
    cost_model: ExperimentPlanningCostModelRequest
    budget: ExperimentPlanningBudgetRequest
    seed: int
    worker_count: int
    failure_policy: Literal["continue_candidate_failures", "fail_fast"]
    created_at: str


class ExperimentLaunchRequest(ExperimentPlanningRequest):
    """Canonical planning document plus the operator-confirmed plan identity."""

    confirmed_plan_hash: str


class ExperimentPreflightCheckResponse(BaseModel):
    """Transport view of one deterministic planning gate result."""

    model_config = ConfigDict(frozen=True)

    rule_id: str
    outcome: str
    code: str | None
    reason: str | None
    remediation: str | None
    observed: dict[str, JsonValue]
    policy: dict[str, JsonValue]


class ExperimentPreflightResponse(BaseModel):
    """Complete operator confirmation surface returned by read-only preflight."""

    model_config = ConfigDict(frozen=True)

    status: str
    plan_hash: str | None
    checks: list[ExperimentPreflightCheckResponse]
    candidate_count: int
    planned_fold_count: int
    budget_run_count: int
    estimated_trading_sessions: int
    estimated_disk_bytes: int
    eligible_month_count: int
    isolation_width_sessions: int


class ExperimentLaunchResponse(BaseModel):
    """Durable server truth returned after launch or exact replay."""

    model_config = ConfigDict(frozen=True)

    experiment_id: str
    status: str
    queue_ordinal: int
    revision: int
    candidate_count: int
    fold_count: int
    plan_hash: str


class ExperimentCandidateResponse(BaseModel):
    """API view of one immutable experiment candidate."""

    model_config = ConfigDict(frozen=True)

    candidate_id: str
    ordinal: int
    is_baseline: bool
    parameters: dict[str, Any]


class ExperimentFoldResponse(BaseModel):
    """API view of one persisted fold specification and projection."""

    model_config = ConfigDict(frozen=True)

    candidate_id: str
    fold_id: str
    ordinal: int
    role: str
    status: str
    train_start: date | None
    train_end: date | None
    test_start: date
    test_end: date
    purge_sessions: int
    embargo_sessions: int
    claim_owner_token: str | None
    revision: int
    updated_at: datetime


class ExperimentDetailResponse(BaseModel):
    """API view of one durable experiment's current server truth."""

    model_config = ConfigDict(frozen=True)

    experiment_id: str
    research_cycle_id: str
    research_cycle_hash: str
    strategy_version: str
    strategy_spec_hash: str
    snapshot_id: str
    status: str
    desired_state: str
    stage: str
    failure_code: str | None
    queue_ordinal: int | None
    revision: int
    created_at: datetime
    updated_at: datetime
    seed: int
    worker_count: int
    failure_policy: str
    candidate_limit: int
    fold_run_limit: int
    fold_protocol_id: str
    fold_protocol_version: int
    fold_protocol_hash: str
    candidate_count: int
    fold_count: int
    candidates: list[ExperimentCandidateResponse]
    folds: list[ExperimentFoldResponse]


class ExperimentGateResponse(BaseModel):
    """API view of one append-only gate evaluation."""

    model_config = ConfigDict(frozen=True)

    evaluation_id: str
    experiment_id: str
    candidate_id: str | None
    fold_id: str | None
    attempt_id: str | None
    rule_id: str
    policy_version: str
    layer: str
    outcome: str
    observed: Any
    policy: Any
    artifact_id: str | None
    payload_hash: str
    evaluated_at: datetime


class ExperimentArtifactResponse(BaseModel):
    """API view of one immutable indexed experiment artifact."""

    model_config = ConfigDict(frozen=True)

    artifact_id: str
    experiment_id: str
    candidate_id: str | None
    fold_id: str | None
    attempt_id: str | None
    artifact_kind: str
    relative_path: str
    content_hash: str
    schema_hash: str
    row_count: int
    byte_size: int
    reproduction_fingerprint: str
    manifest: Any
    is_pinned: bool
    pinned_at: datetime | None
    created_at: datetime
    revision: int


class ExperimentSelectionEvidenceResponse(BaseModel):
    """API view of one experiment's verified selection-evidence ledger."""

    model_config = ConfigDict(frozen=True)

    artifact_id: str
    experiment_id: str
    content_hash: str
    byte_size: int
    is_pinned: bool
    created_at: datetime
    payload: Any


class ExperimentComparisonResponse(BaseModel):
    """API view of one experiment's candidate comparison projection."""

    model_config = ConfigDict(frozen=True)

    experiment_id: str
    payload_hash: str
    revision: int
    payload: Any


class ReviewGateOutcomeResponse(BaseModel):
    """API view of one gate rule's identity and outcome in a review packet."""

    model_config = ConfigDict(frozen=True)

    rule_id: str
    layer: str
    outcome: str


class ReviewSelectionTraceRefResponse(BaseModel):
    """API view of one verified selection-trace artifact reference."""

    model_config = ConfigDict(frozen=True)

    artifact_kind: str
    artifact_id: str
    content_hash: str


class ReviewExposureWeightResponse(BaseModel):
    """One canonical exposure dimension and aggregate selected weight."""

    model_config = ConfigDict(frozen=True)

    key: str
    weight: float


class ReviewSelectionExposureResponse(BaseModel):
    """Explicit stock exposure or ETF not-applicable review evidence."""

    model_config = ConfigDict(frozen=True)

    applicability: str
    lane: str
    industry_weights: list[ReviewExposureWeightResponse]
    size_bucket_weights: list[ReviewExposureWeightResponse]
    artifact_refs: list[ReviewSelectionTraceRefResponse]


class ExperimentReviewPacketResponse(BaseModel):
    """API view of one immutable promotion review packet."""

    model_config = ConfigDict(frozen=True)

    experiment_id: str
    candidate_id: str | None
    bundle_hash: str
    hard_review_blocked: bool
    gate_outcomes: list[ReviewGateOutcomeResponse]
    schema_version: int
    fold_ids: list[str]
    attempt_ids: list[str]
    spec_hash: str
    resolved_spec_hash: str
    parameter_hash: str
    snapshot_hash: str
    registry_hash: str
    objective_payload_hash: str
    comparison_payload_hash: str | None
    r1_impact_payload_hash: str | None
    selection_evidence_artifact_id: str | None
    holdout_claim_id: str | None
    candidate_rationale: str
    selection_trace_artifact_refs: list[ReviewSelectionTraceRefResponse]
    selection_exposure: ReviewSelectionExposureResponse | None


class ExperimentSummaryResponse(BaseModel):
    """API view of one experiment root in list views (no candidate/fold expansion)."""

    experiment_id: str
    status: str
    desired_state: str
    stage: str
    failure_code: str | None
    queue_ordinal: int | None
    revision: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(frozen=True)


class NodeDescriptorResponse(BaseModel):
    """API view of one immutable strategy pipeline node descriptor."""

    node_type: str
    version: str
    category: str
    display_name: str
    implementation_key: str
    config_schema: dict[str, str]
    default_config: dict[str, Any]
    required_datasets: list[str]
    capability_tags: list[str]
    deterministic: bool

    model_config = ConfigDict(strict=True, extra="ignore")


class FactorDescriptorResponse(BaseModel):
    """API view of one governed core-factor descriptor."""

    factor_id: str
    resolved_payload: dict[str, Any]

    model_config = ConfigDict(strict=True, extra="ignore")


class FactorDiagnosticsResponse(BaseModel):
    """Typed immutable factor diagnostics and their complete read identity."""

    model_config = ConfigDict(frozen=True)

    factor_id: str
    snapshot_id: str
    snapshot_hash: str
    registry_hash: str
    start_date: date
    end_date: date
    provenance: dict[str, Any]
    metrics: dict[str, Any]
    artifact_id: str
    content_hash: str


class CandidateSelectionRequest(BaseModel):
    """Body for one durable server-side promotion preselection."""

    model_config = ConfigDict(frozen=True)

    candidate_id: str
    rationale: str
    comparison_payload_hash: str
    expected_revision: int


class CandidateSelectionReceiptResponse(BaseModel):
    """API projection of one append-only candidate preselection event."""

    model_config = ConfigDict(frozen=True)

    selection_id: str
    experiment_id: str
    candidate_id: str
    comparison_payload_hash: str
    candidate_evidence_artifact_id: str
    candidate_evidence_content_hash: str
    selection_evidence_content_hash: str
    revision: int
    event_id: str
    occurred_at: datetime


class HoldoutSelectionReasonRequest(BaseModel):
    """Operator-authored reason bound into the immutable holdout claim."""

    model_config = ConfigDict(frozen=True)

    code: str
    summary: str


class HoldoutEvaluationRequest(BaseModel):
    """Body for the one-shot holdout claim of a persisted preselection."""

    model_config = ConfigDict(frozen=True)

    candidate_id: str
    selection_id: str
    expected_selection_evidence_hash: str
    expected_candidate_evidence_content_hash: str
    expected_revision: int
    operator_confirmation: str
    selection_reason: HoldoutSelectionReasonRequest


class HoldoutEvaluationReceiptResponse(BaseModel):
    """API projection of one committed or exactly replayed holdout claim."""

    model_config = ConfigDict(frozen=True)

    claim_id: str
    selection_id: str
    experiment_id: str
    candidate_id: str
    state: Literal["claimed"]
    fold_id: str
    logical_run_id: str
    reproduction_fingerprint: str
    claim_payload_hash: str
    selection_evidence_content_hash: str
    candidate_evidence_content_hash: str
    revision: int
    event_id: str
    occurred_at: datetime


class CandidateSelectionEventResponse(BaseModel):
    """One candidate selection event from an immutable candidate bundle."""

    model_config = ConfigDict(frozen=True)

    validation_fold_ordinal: int
    fold_id: str
    trade_date: str
    instrument_id: int | str
    score: float | None
    rank: int
    selected: bool
    evidence_hash: str


class CandidateExclusionEventResponse(BaseModel):
    """One candidate exclusion event from an immutable candidate bundle."""

    model_config = ConfigDict(frozen=True)

    validation_fold_ordinal: int
    fold_id: str
    trade_date: str
    instrument_id: int | str
    stage: str
    reason_code: str
    message: str | None
    evidence_hash: str


class CandidateFactorContributionResponse(BaseModel):
    """One factor contribution bound to a fold decision and evidence hash."""

    model_config = ConfigDict(frozen=True)

    validation_fold_ordinal: int
    fold_id: str
    trade_date: str
    instrument_id: int | str
    factor_id: str
    contribution: float | None
    rank: int | None
    selected: bool | None
    evidence_hash: str


class CandidateSelectionPageResponse(BaseModel):
    """Typed selections page over one immutable candidate bundle."""

    model_config = ConfigDict(frozen=True)

    candidate_id: str
    experiment_id: str
    artifact_id: str
    content_hash: str
    items: list[CandidateSelectionEventResponse]
    next_cursor: str | None


class CandidateExclusionPageResponse(BaseModel):
    """Typed exclusions page over one immutable candidate bundle."""

    model_config = ConfigDict(frozen=True)

    candidate_id: str
    experiment_id: str
    artifact_id: str
    content_hash: str
    items: list[CandidateExclusionEventResponse]
    next_cursor: str | None


class CandidateFactorContributionPageResponse(BaseModel):
    """Typed contribution page over one immutable candidate bundle."""

    model_config = ConfigDict(frozen=True)

    candidate_id: str
    experiment_id: str
    artifact_id: str
    content_hash: str
    items: list[CandidateFactorContributionResponse]
    next_cursor: str | None


class ExperimentControlRequest(BaseModel):
    """Body for one revision-fenced experiment control action (pause/cancel/resume)."""

    model_config = ConfigDict(frozen=True)

    expected_revision: int


class ExperimentRetryFoldRequest(BaseModel):
    """
    Body for one revision-fenced terminal fold retry.

    expected_revision is the fold projection revision (not experiment revision).
    """

    model_config = ConfigDict(frozen=True)

    candidate_id: str
    fold_id: str
    expected_revision: int


class ExperimentControlReceiptResponse(BaseModel):
    """API view of one durable control receipt (CAS post-state)."""

    model_config = ConfigDict(frozen=True)

    experiment_id: str
    status: str
    desired_state: str
    revision: int
    occurred_at: datetime
    live_run_ids: list[str]
